x/wasm/wasm_snapshotter.go

```diff
@@ -0,0 +1,150 @@
+package wasm
+
+import (
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+
+	snapshot "github.com/cosmos/cosmos-sdk/snapshots/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	protoio "github.com/gogo/protobuf/io"
+)
+
+/*
+API to implement:
+
+// Snapshotter is something that can create and restore snapshots, consisting of streamed binary
+// chunks - all of which must be read from the channel and closed. If an unsupported format is
+// given, it must return ErrUnknownFormat (possibly wrapped with fmt.Errorf).
+type Snapshotter interface {
+	// Snapshot writes snapshot items into the protobuf writer.
+	Snapshot(height uint64, protoWriter protoio.Writer) error
+
+	// Restore restores a state snapshot from the protobuf items read from the reader.
+	// If the ready channel is non-nil, it returns a ready signal (by being closed) once the
+	// restorer is ready to accept chunks.
+	Restore(height uint64, format uint32, protoReader protoio.Reader) (SnapshotItem, error)
+}
+
+// ExtensionSnapshotter is an extension Snapshotter that is appended to the snapshot stream.
+// ExtensionSnapshotter has an unique name and manages it's own internal formats.
+type ExtensionSnapshotter interface {
+	Snapshotter
+
+	// SnapshotName returns the name of snapshotter, it should be unique in the manager.
+	SnapshotName() string
+
+	// SnapshotFormat returns the default format the extension snapshotter use to encode the
+	// payloads when taking a snapshot.
+	// It's defined within the extension, different from the global format for the whole state-sync snapshot.
+	SnapshotFormat() uint32
+
+	// SupportedFormats returns a list of formats it can restore from.
+	SupportedFormats() []uint32
+}
+*/
+
+type WasmSnapshotter struct {
+	wasmDirectory string
+}
+
+func NewWasmSnapshotter(wasmDirectory string) *WasmSnapshotter {
+	return &WasmSnapshotter{
+		wasmDirectory,
+	}
+}
+
+func (ws *WasmSnapshotter) SnapshotName() string {
+	return "WASM Files Snapshot"
+}
+
+func (ws *WasmSnapshotter) SnapshotFormat() uint32 {
+	return 1
+}
+
+func (ws *WasmSnapshotter) SupportedFormats() []uint32 {
+	return []uint32{1}
+}
+
+var wasmFileNameRegex = regexp.MustCompile(`^[a-f0-9]{64}$`)
```

Comment:
Those file names can change any time. I'm actually thinking about adding a .wasm file extension for node operator friendlyness.

Reply:
This is a good point, but for a given wasmd version, there will be only one storage format (and realistically, changing this will be a major state-breaking change needing an internal wasmvm migration, so unlikely to happen, or at least very infrequently).

If it does change, we make a new wasmd (eg. 2.0) with snapshot format 2. Upon receiving that the data is parsed in a different way (eg the new internal format). The snapshotter itself (they one reading), will just need to implement the new format when we break this internal storage format.

Reply:
That said, I do agree with using the GetCode and Create calls

Reply:
I agree that it's better to do it through the proper API, and let future migrations less to worry about WRT snapshotting.

Reply:
This also "solves" the "wasm/wasm/state/wasm` issue here.